ulib/experimental/Steel.Effect.fsti

```diff
@@ -374,33 +209,48 @@ let bind_pure_steel__ens (#a:Type) (#b:Type)
 
 val bind_pure_steel_ (a:Type) (b:Type)
   (#[@@@ framing_implicit] wp:pure_wp a)
+  (#framed:eqtype_as_type bool)
   (#[@@@ framing_implicit] pre:pre_t) (#[@@@ framing_implicit] post:post_t b)
   (#[@@@ framing_implicit] req:a -> req_t pre) (#[@@@ framing_implicit] ens:a -> ens_t pre b post)
-  (f:eqtype_as_type unit -> PURE a wp) (g:(x:a -> repr b pre post (req x) (ens x)))
+  (f:eqtype_as_type unit -> PURE a wp) (g:(x:a -> repr b framed pre post (req x) (ens x)))
 : repr b
+    framed
     pre
     post
     (bind_pure_steel__req wp req)
     (bind_pure_steel__ens wp ens)
 
-polymonadic_bind (PURE, SteelF) |> SteelF = bind_pure_steel_
+polymonadic_bind (PURE, SteelBase) |> SteelBase = bind_pure_steel_
 
-polymonadic_bind (PURE, Steel) |> Steel = bind_pure_steel_
+// AF: The lift below works, but the framing tactic currently requires annotations
+// when scope restriction occurs (i.e. during returns). See for instance the comment
+// on test7 in Steel.FramingTestSuite. When a lift is available, returns are not inserted,
+// which leads to errors
 
+// unfold
+// let lift_pure_steel__req (#a:Type) (wp:pure_wp a)
+// : req_t emp
+// = fun m -> as_requires wp /\ True
 
-(*
-//  * subcomp relation from SteelF to Steel
-//  *)
+// unfold
+// let lift_pure_steel__ens (#a:Type) (wp:pure_wp a)
+// : ens_t emp a (fun _ -> emp)
+// = fun m0 r m1 -> as_requires wp /\ as_ensures wp r
 
-polymonadic_subcomp SteelF <: Steel = subcomp
+// val lift_pure_steel
+//   (a:Type)
+//   (#[@@@ framing_implicit] wp:pure_wp a)
+//   (f:eqtype_as_type unit -> PURE a wp)
+//   : repr a false emp (fun _ -> emp) (lift_pure_steel__req wp) (lift_pure_steel__ens wp)
 
+// sub_effect PURE ~> SteelBase = lift_pure_steel
 
 (*
 //  * Annotations without the req and ens
 //  *)
 
 effect SteelT (a:Type) (pre:pre_t) (post:post_t a) =
-  Steel a pre post (fun _ -> True) (fun _ _ _ -> True)
+  SteelBase a false pre post (fun _ -> True) (fun _ _ _ -> True)
```

Comment:
For readability you could define it as `Steel a pre post (fun _ -> True) (fun _ _ _ -> True)`.

Reply:
Done